systemtest/src/main/java/io/strimzi/systemtest/parallel/SuiteThreadController.java

```diff
@@ -103,8 +119,68 @@ public synchronized void waitUntilEntryIsOpen(ExtensionContext extensionContext)
+     * and thus many threads can start execute test suites which could potentially destroy cluster. This is mechanism,
+     * which will all additional threads (i.e., not needed) put into waiting room. After one of the
+     * {@link io.strimzi.systemtest.annotations.ParallelSuite} is done with execution we release
+     * ({@link #notifyParallelSuiteToAllowExecution(ExtensionContext)} one test suite setting {@code isParallelSuiteReleased} flag.
+     * This ensures that only one test suite will continue with execution and others will still wait.
+     *
+     * @param extensionContext Extension context for test suite name
+     */
+    public void waitUntilAllowedNumberTestSuitesInParallel(ExtensionContext extensionContext) {
+        final String testSuiteToWait = extensionContext.getRequiredTestClass().getSimpleName();
+        waitingTestSuites.add(testSuiteToWait);
+
+        while (!isRunningAllowedNumberTestSuitesInParallel()) {
+            LOGGER.debug("{} is waiting to proceed with execution but current thread exceed maximum " +
+                "of allowed test suites in parallel. ({}/{})",
+                testSuiteToWait,
+                runningTestSuitesInParallelCount.get(),
+                maxTestSuitesInParallel);
+
+            try {
+                Thread.currentThread().sleep(STARTING_DELAY);
+            } catch (InterruptedException e) {
+                throw new RuntimeException(e);
+            }
+
+            // release and lock again
+            if (isParallelSuiteReleased.get()) {
+                // lock
+                isParallelSuiteReleased.set(false);
+                waitingTestSuites.remove(testSuiteToWait);
+                break;
+            }
+        }
+        LOGGER.debug("{} suite now can proceed its execution", testSuiteToWait);
+    }
+
+    public void incrementParallelSuiteCounter() {
+        runningTestSuitesInParallelCount.set(runningTestSuitesInParallelCount.incrementAndGet());
+    }
+
     public void decrementCounter() {
-        runningTestSuitesInParallelCount.decrementAndGet();
+        runningTestSuitesInParallelCount.set(runningTestSuitesInParallelCount.decrementAndGet());
+    }
+
+    /**
+     * Notifies one of the {@link io.strimzi.systemtest.annotations.ParallelSuite} to continue its execution. Specifically
+     * {@code waitingTestSuites}, which are waiting because {@link java.util.concurrent.ForkJoinPool} spawns
+     * additional threads, which exceed parallelism limit. This ensures that {@link io.strimzi.systemtest.annotations.ParallelSuite}
+     * will not deadlock.
+     *
+     * @param extensionContext extension context for identifying, which test suite notifies.
+     */
+    public void notifyParallelSuiteToAllowExecution(ExtensionContext extensionContext) {
+        LOGGER.info("{} - Notifies waiting test suites:{} to and randomly select one to start execution", extensionContext.getRequiredTestClass().getSimpleName(), waitingTestSuites.toString());
```

Comment:
```suggestion
        LOGGER.debug("[{}] - Notifies waiting test suites:{} to and randomly select one to start execution", extensionContext.getRequiredTestClass().getSimpleName(), waitingTestSuites.toString());
```

Are you sure this won't print the same class name every time?

Reply:
I am 100% sure that it will not print the same class. This method is called in @AfterAll, which means that each test suite will have a different ExtensionContext object and thus a different name.